api/src/utils/progress.ts

```diff
@@ -0,0 +1,22 @@
+export type ProgressTimestamp = number | { timestamp: number } | null;
+
+export const getCalendar = (
+  progressTimestamps?: ProgressTimestamp[]
+): Record<string, 1> => {
+  const calendar: Record<string, 1> = {};
+
+  progressTimestamps?.forEach(progress => {
+    if (progress === null) return;
+    if (typeof progress === 'number') {
+      calendar[Math.floor(progress / 1000)] = 1;
+    } else {
+      calendar[Math.floor(progress.timestamp / 1000)] = 1;
+    }
+  });
+
+  return calendar;
+};
+
+export const getPoints = (progressTimestamps?: ProgressTimestamp[]): number => {
+  return progressTimestamps?.length ?? 1;
+};
```

Comment:
Can you give me a walkthrough of why they need to be `undefined`? will they be used without TypeScript warning about the arguments?

Reply:
Are you asking why the two functions need to be able to handle `undefined` arguments?  It's not 100% necessary, but it seemed cleaner than conditionally call the functions if `progressTimestamps` is defined.

We know (from Shaun's work defining the schema) that not all user documents have a `progressTimestamps` field, so we have to deal with the documents where it's missing.

Having said that, I think prisma will return `null`, not `undefined`, but I have to check.

> will they be used without TypeScript warning about the arguments?

I didn't quite understand this question. Could you elaborate?

Reply:
I haven't thought of the user documents and their lack of `progressTimestamps`, now this makes sense.

because of me not thinking about the user documents, I couldn't find a reason to not call it conditionally, and in attempt to understand why would it be `undefined`, I needed a walkthrough for why, thank you. 👍 

---

> I didn't quite understand this question. Could you elaborate?

I have thought that, we will need conditional call, because it didn't make sense that a call will be made to get a point or date, without the progress being made, so `progressTimestamps` will never be `undefined`.